Paraswap-Agents/delegate-calls/src/agent.ts

```diff
@@ -0,0 +1,100 @@
+  "event RoleRevoked(bytes32 indexed role, address indexed account, address indexed sender)",
+];
+
+// Checks if `logicAddr` has the role `ROUTER_ROLE` in the contract `paraAddr`
+const checkRouterRole = async (
+  paraAddr: string,
+  logicAddr: string,
+  block: number | string
+): Promise<boolean | undefined> => {
+  let isRouterRole;
+  // If the contract exists in the cache then return data
+  if (cache.has(logicAddr)) {
+    isRouterRole = cache.get(logicAddr);
+  // Otherwise the cache does not contain data for `logicAddr`
+  } else {
+    try {
+      // Check on-chain data to see if `logicAddr` has the correct role
+      const contractInterface = new ethers.Contract(paraAddr, PARA_ABI, getEthersProvider());
+      isRouterRole = await contractInterface.hasRole(ROUTER_ROLE, logicAddr, { blocktag: block });
+      cache.set(logicAddr, isRouterRole);
+    } catch (error) {
+      // If there is as error set `isRouterRole` to undefined
+      // A special finding to indicate that the on-chain call failed is generated when `isRouterRole` is undefined
+      // This is not saved to cache in case it was a short term issue
+      isRouterRole = undefined;
+    }
+  }
+  return isRouterRole;
+};
+
+export const provideHandleTransaction = (address: string, checkRouterRole: any): HandleTransaction => {
+  return async (tx: TransactionEvent): Promise<Finding[]> => {
+    // Setup the findings array
+    const findings: Finding[] = [];
+
+    // Check if any contracts have had roles granted and update `cache` if needed
+    const roleGrants = tx.filterLog(PARA_ABI[1], address);
+    roleGrants.forEach((roleGrant) => {
+      // If ROUTER_ROLE was granted to an account that is inside the cache then update the cache
+      if (roleGrant.args.role == ROUTER_ROLE && cache.has(roleGrant.args.account)) {
+        cache.set(roleGrant.args.account, true);
+      }
+    });
+
+    // Check if any contracts have had roles revoked and update `cache` if needed
+    const roleRevokes = tx.filterLog(PARA_ABI[2], address);
+    roleRevokes.forEach((roleRevoke) => {
+      // If ROUTER_ROLE was removed from an account that is inside the cache then update the cache
+      if (roleRevoke.args.role == ROUTER_ROLE && cache.has(roleRevoke.args.account)) {
+        cache.set(roleRevoke.args.account, false);
+      }
+    });
+
+    // For each trace in the transaction
+    await Promise.all(
+      tx.traces.map(async (trace) => {
+        // If the calltype is "delegatecall" and the from address is `address`
+        if (trace.action.callType == "delegatecall" && trace.action.from == address) {
+          // Check if the logic contract for the delegatecall has the role ` ROUTER_ROLE`
+          const hasRouterRole = await checkRouterRole(address, trace.action.to, tx.blockNumber);
+          // Create a finding
+          findings.push(createFinding(hasRouterRole, trace.action.to));
+        }
+      })
+    );
```

Comment:
Since the `AugustusSwapper` contract is a proxy that `delegatecall`s to other contracts for most of its functionality, I think it would be more helpful to be selective under what conditions Findings are created. Specifically when the logic contract **has** the `ROUTER_ROLE`.

Perhaps filtering for specific function calls in the logic contract (e.g. `multiswap` and `megaswap` per the Paraswap Notion page)? To be even more selective, perhaps, using `multiswap` and `megaswap` as examples, only when `toAmount` passes a certain threshold.

Reply:
That is right, we are going to deliver detection for only `multiswap` and `megaswap` calls + the `ROUTER_ROLE` monitoring. So please add the findings for each time that functions are called.